opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -550,19 +561,18 @@ def get_span_context(self):
         return self.context
 
     def set_attribute(self, key: str, value: types.AttributeValue) -> None:
-        with self._lock:
-            if not self.is_recording():
-                return
-            has_ended = self.end_time is not None
-        if has_ended:
-            logger.warning("Setting attribute on ended span.")
+        if not _is_valid_attribute_value(value):
             return
 
         if not key:
             logger.warning("invalid key (empty or null)")
             return
 
-        if _is_valid_attribute_value(value):
+        with self._lock:
```

Comment:
We aren't using `@_check_span_ended` here?

Reply:
no, i wanted to keep the `_is_valid_attribute_value` outside of the lock